v2/pkg/engine/resolve/loader.go

```diff
@@ -1599,19 +1599,27 @@ func (l *Loader) executeSourceLoad(ctx context.Context, fetchItem *FetchItem, so
 		}
 	}
 	if l.propagateFetchReasons && !IsIntrospectionDataSource(res.ds.ID) {
-		fetchReasons := fetchItem.Fetch.FetchReasons()
+		fetchReasons := fetchItem.Fetch.GetInfo().FetchReasons
 		if len(fetchReasons) > 0 {
-			var encoded []byte
-			encoded, res.err = json.Marshal(fetchReasons)
-			if res.err != nil {
-				res.err = errors.WithStack(res.err)
-				return
+			// Collect only those FetchReasons that were requested via directives.
+			indices := fetchItem.Fetch.GetInfo().RequireFetchReason
+			filteredReasons := make([]FetchReason, 0, len(indices))
+			for _, idx := range indices {
+				filteredReasons = append(filteredReasons, fetchReasons[idx])
 			}
-			// We expect that body.extensions is an object
-			input, res.err = jsonparser.Set(input, encoded, "body", "extensions", "fetch_reasons")
-			if res.err != nil {
-				res.err = errors.WithStack(res.err)
-				return
+			if len(filteredReasons) > 0 {
+				var encoded []byte
+				encoded, res.err = json.Marshal(filteredReasons)
+				if res.err != nil {
+					res.err = errors.WithStack(res.err)
+					return
+				}
+				// We expect that body.extensions is an object
+				input, res.err = jsonparser.Set(input, encoded, "body", "extensions", "fetch_reasons")
```

Comment:
I don't want to be annoying but I think this approach is wrong. We have an existing architecture to render the input. Now we add post processing to extend it. I understand that this makes things simpler. However, this is not a universal solution. E.g. if we want to add fetch reasons to gRPC messages, we will now have to parse the input and extract the information again. Wouldn't it be cleaner if we make it part of the input template and just render it into the input? No post processing needed. And if we want to add fetch reasons to gRPC, we can update the gRPC templating pipeline and provide fetch reasons statically typed instead of just in input. my 2 cents

Reply:
I believe it should not be a part of this issue, as the code for propagating arbitrary request extensions was already there

Could be part of a bigger refactoring of making input an object instead of a string

Reply:
@jens, I have created corresponding issue, please mark as resolved :)